utils/pattern/src/pattern/mod.rs

```diff
@@ -0,0 +1,185 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+mod error;
+
+use crate::parser::error::ParserError;
+use crate::token::PatternToken;
+use crate::{
+    interpolator::{InterpolatedKind, Interpolator, InterpolatorError},
+    parser::{Parser, ParserOptions},
+    replacement::ReplacementProvider,
+};
+pub use error::PatternError;
+use std::{
+    convert::{TryFrom, TryInto},
+    fmt::{Debug, Display, Write},
+    str::FromStr,
+};
+
+/// `Pattern` stores the result of parsing operation as a vector
+/// of [`PatternToken`] elements.
+///
+/// # Type parameters
+///
+/// - `P`: The type of the placeholder used as a key for the [`ReplacementProvider`].
+///
+/// # Lifetimes
+///
+/// - `p`: The life time of an input string slice to be parsed.
+///
+/// [`ReplacementProvider`]: crate::ReplacementProvider
+pub struct Pattern<'s, P>(pub(crate) Vec<PatternToken<'s, P>>);
```

Comment:
question: as Shane says elsewhere, is there a reason this can not just be a slice?

Reply:
So, I think interpolator can take a slide (and I added it via `Deref`). If we make Pattern be a wrapper over slice, then I think it loses any value of having it over just `&[PatternToken]` at all.
I see `Pattern` as potentially becoming the struct that you store in Provider struct, so maybe `Cow<[PatternToken]>`? I'd also love to avoid having to bikeshed that here and consider that when we'll be using `icu_pattern`.

Reply:
Let's get this checked in and then make it be backed by ZeroVec.